### PR TITLE
[7/n][dagster-airbyte] Implement get_asset_spec in DagsterAirbyteTranslator

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

This PR implements `get_asset_spec` in the `DagsterAirbyteTranslator` using the props object created in the previous PR.

## How I Tested These Changes

Additional unit tests with BK

### CODE REVIEW

**Comment by maximearmstrong** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26264?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26264** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26264?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26264?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26257** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26257?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26253** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26253?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26242** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26242?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26241** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26205** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26204** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py [hunk 1]
```diff
@@ -140,4 +142,29 @@ class DagsterAirbyteTranslator:
...
+        )
+
+        metadata = {
+            **TableMetadataSet(
+                column_schema=column_schema,
+                table_name=fully_qualified_table_name,
+            ),
+            "connection_id": props.connection_id,
+            "connection_name": props.connection_name,
+            "stream_prefix": props.stream_prefix,
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py:163]: we should be using real english for metadata keys if they're intended for human usage; and if they're not, we should be prefixing them with dagster-airbyte/...
Either way, create constants for these

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py [hunk 2]
```diff
@@ -140,4 +142,29 @@ class DagsterAirbyteTranslator:
...
+        table_schema_props = props.json_schema.get("properties", {}) or props.json_schema.get(
+            "items", {}
+        ).get("properties", {})
+        column_schema = generate_table_schema(table_schema_props)
+
+        fully_qualified_table_name = (
+            f"{props.database}.{props.schema}.{props.stream_name}"
+            if props.database and props.schema
+            else None
+        )
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py:154]: what's the context where this is actually none?

Given that this is entirely part of props maybe a property/method.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_translator.py [hunk 3]
```diff
@@ -37,3 +41,70 @@ def test_airbyte_workspace_data_to_table_props(
...
+
+    translator = DagsterAirbyteTranslator()
+    asset_spec = translator.get_asset_spec(first_table_props)
+
+    assert asset_spec.key.path == ["test_prefix_test_stream"]
+    assert asset_spec.metadata["connection_id"] == TEST_CONNECTION_ID
+    assert asset_spec.metadata["connection_name"] == TEST_CONNECTION_NAME
+    assert asset_spec.metadata["stream_prefix"] == TEST_STREAM_PREFIX
+    assert (
+        asset_spec.metadata["dagster/table_name"]
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_translator.py:69]: use constant

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_translator.py [hunk 4]
```diff
@@ -37,3 +41,70 @@ def test_airbyte_workspace_data_to_table_props(
...
+        == f"{TEST_DESTINATION_DATABASE}.{TEST_DESTINATION_SCHEMA}.{TEST_STREAM_NAME}"
+    )
+    assert asset_spec.metadata["dagster/column_schema"] == generate_table_schema(
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_translator.py:72]: use constant

**Review by dpeng817 — APPROVED** [review]

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py [hunk 5]
```diff
@@ -143,4 +145,29 @@ class DagsterAirbyteTranslator:
 
     def get_asset_spec(self, props: AirbyteConnectionTableProps) -> AssetSpec:
         """Get the AssetSpec for a table synced by an Airbyte connection."""
-        raise NotImplementedError()
+        table_schema_props = props.json_schema.get("properties", {}) or props.json_schema.get(
+            "items", {}
+        ).get("properties", {})
+        column_schema = generate_table_schema(table_schema_props)
```
[the remark this answers is not in the document]
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py]: Done in [4648e5c](https://github.com/dagster-io/dagster/pull/26264/commits/4648e5c376d0511e53e22f40496286e73df97198)

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py (hunk 1 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/translator.py]: That's fair - I implemented a `AirbyteMetadataSet` and used `.extract(metadata)` in tests in [9e904e7](https://github.com/dagster-io/dagster/pull/26264/commits/9e904e7bc7a83bd693699c9c9363ac45340fe62f)

**Pull request merged**